lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+				//TODO: understand what are these -> other analysis? so should I keep them ? I think I should
+				solver->load<mlir::dataflow::DeadCodeAnalysis>();
+				solver->load<mlir::dataflow::SparseConstantPropagation>();
+				analysis = solver->load<ConstraintsAnalysis>(symbolTable);
+				auto top = fun;
+				llvm::outs()<<"Initialize analysis from operation: ";
+				top->print(llvm::outs());
+				auto res = solver->initializeAndRun(top);
+			}
+			
+			void printRangesFound(mlir::Operation* op){
+				llvm::outs()<<"Printing operation: ";
+				op->print(llvm::outs());
+				llvm::outs()<<"\n";
+				this->analysis->printRanges(op);
+			}
+
+			const llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange>& getOperationLattice(mlir::Operation* op){
+				return this->analysis->getUnderlyingLattice(op);
+			}
+
+			private:
+			mlir::SymbolTableCollection symbolTable;
+			mlir::rlc::FlatFunctionOp function;
+			DataFlowConfig config;
+			std::unique_ptr<DataFlowSolver> solver;
+			ConstraintsAnalysis* analysis;
+		};
+
+	}	 // namespace
+
+
+#define GEN_PASS_DEF_CONSTRAINTSANALYSISPASS
+#include "rlc/dialect/Passes.inc"
+
+	struct ConstraintsAnalysisPass
+			: impl::ConstraintsAnalysisPassBase<ConstraintsAnalysisPass>
+	{
+		using impl::ConstraintsAnalysisPassBase<
+				ConstraintsAnalysisPass>::ConstraintsAnalysisPassBase;
+
+		void runOnOperation() override
+		{
+			//Perform the analysis only if the function returns a boolean
+			for (auto fun : getOperation().getOps<mlir::rlc::FlatFunctionOp>())
+			{
+				
+				if(not mlir::isa<mlir::rlc::BoolType>(fun.getResult().getType().getResult(0)) )
+					continue;
+
+				if(fun.getBlocks().size()<=2)
+					continue;
+
+				Constraints con(fun);
+				std::cout<<"In theory I should have run the analysis"<<std::endl;
+				// Here I append to the function the ranges I found
+				// TODO: understand why the analysis is done backwards but the blocks are traversed forwards
+				// https://mlir.llvm.org/doxygen/DenseAnalysis_8cpp_source.html#l00387 <- I am just stupid, here it goes through the predecessors
+				// Blocks have a getPredecessors() method
+				auto block=fun.getRegion(0).front().getNextNode()->getNextNode();
+				con.printRangesFound(&block->back());
+
+				auto lattice = con.getOperationLattice(&block->back());
+
+				// Iterate through the arguments -> are the arguments of the first block
+				for(auto arg: fun.getRegion(0).front().getArguments()){
+					auto casted=mlir::dyn_cast<mlir::Value>(arg);
+					for(auto pair : lattice){
+						if(pair.first.first==casted){
+							std::string to_write;
+							auto stream=llvm::raw_string_ostream(to_write);
+							arg.printAsOperand(stream,mlir::OpPrintingFlags());
+							stream.flush();
+							// Add range as a function argument
+							fun->setAttr("rlc.attr."+to_write+"_branch_"+std::to_string(pair.first.second)+"_MIN",mlir::IntegerAttr::get(mlir::IntegerType::get(fun->getContext(),32),pair.second.getMin()));
```

Comment:
not sure what is going on here, can't we jut do?
```cpp
setAttr(fun.getArgumentName()[index], "(" + to_string(range.min()) + ","+ to_string(range.max()) + ")");
```

Reply:
hmm yes now that I played around a bit there is a simpler way to write those arguments.

The only "problem" is that the method signature is not exactly the same but it should work the same way.

And also I was adding the arguments as mlir::IntegerAttr but if we want them as simple strings then its also ok.

Reply:
we are only doing this to test, so in the end they all end up as strings anyway